pkg/e2e/fake_metrics_server_test.go

```diff
@@ -0,0 +1,315 @@
+		Name:       "PROJECT/1234",
+		TimeSeries: timeSeries,
+	}
+	fms.CreateTimeSeries(context.TODO(), createTimeSeriesRequest)
+
+	// these are the subtests
+	tests := []*createTimeSeriesTest{
+		{testName: "TestNil"},
+		{
+			testName: "TestExceedMaxTimeSeriesPerRequest",
+			createTimeSeriesRequest: &monitoringpb.CreateTimeSeriesRequest{
+				Name:       "PROJECT/1234",
+				TimeSeries: []*monitoringpb.TimeSeries{{}, {}},
+			},
+		},
+		{
+			testName: "TestNoTimeSeriesToAdd",
+			createTimeSeriesRequest: &monitoringpb.CreateTimeSeriesRequest{
+				Name: "PROJECT/1234",
+			},
+		},
+		{
+			testName: "TestNoPointInTimeSeriesToAdd",
+			createTimeSeriesRequest: &monitoringpb.CreateTimeSeriesRequest{
+				Name: "PROJECT/1234",
+				TimeSeries: []*monitoringpb.TimeSeries{{
+					Resource: &monitoredrespb.MonitoredResource{
+						Type: "prometheus_target",
+						Labels: map[string]string{
+							"project_id": "example-project",
+							"location":   "europe",
+							"cluster":    "foo-cluster",
+							"namespace":  "",
+							"job":        "job1",
+							"instance":   "instance1",
+						},
+					},
+					Metric: &metricpb.Metric{
+						Type:   "prometheus.googleapis.com/metric1/gauge",
+						Labels: map[string]string{"k1": "v1"},
+					},
+					MetricKind: metricpb.MetricDescriptor_GAUGE,
+					ValueType:  metricpb.MetricDescriptor_DOUBLE,
+				}},
+			},
+		},
+		{
+			testName: "TestAddPointInPast",
+			createTimeSeriesRequest: &monitoringpb.CreateTimeSeriesRequest{
+				Name: "PROJECT/1234",
+				TimeSeries: []*monitoringpb.TimeSeries{{
+					Resource: &monitoredrespb.MonitoredResource{
+						Type: "prometheus_target",
+						Labels: map[string]string{
+							"project_id": "example-project",
+							"location":   "europe",
+							"cluster":    "foo-cluster",
+							"namespace":  "",
+							"job":        "job1",
+							"instance":   "instance1",
+						},
+					},
+					Metric: &metricpb.Metric{
+						Type:   "prometheus.googleapis.com/metric1/gauge",
+						Labels: map[string]string{"k1": "v1"},
+					},
+					MetricKind: metricpb.MetricDescriptor_GAUGE,
+					ValueType:  metricpb.MetricDescriptor_DOUBLE,
+					Points: []*monitoringpb.Point{{
+						Interval: &monitoringpb.TimeInterval{
+							StartTime: &timestamppb.Timestamp{Seconds: 1},
+							EndTime:   &timestamppb.Timestamp{Seconds: 2},
+						},
+						Value: &monitoringpb.TypedValue{
+							Value: &monitoringpb.TypedValue_DoubleValue{DoubleValue: 0.6},
+						},
+					}},
+				}},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.testName, func(t *testing.T) {
+			response, err := fms.CreateTimeSeries(context.TODO(), test.createTimeSeriesRequest)
+			if err == nil && response != nil {
+				t.Errorf("expected an error for %q", test.testName)
+			}
+		})
+	}
+}
+
+func TestCreateTimeSeries(t *testing.T) {
+	fms := NewFakeMetricServer(200)
+	projectName := "PROJECT/1234"
+
+	// these are the subtests
+	tests := []*createTimeSeriesTest{
```

Comment:
I'll suggest change this into one test case structure because each cases depended on previous: new project, existing project, existing project existing time series. I guess most ppl have the assumption that each test case is independent with each other. So if you want dependency to mock a stateful service, make 3 requests in one test case.

Reply:
@andysim3d Should I revert it to it's previous state (ie: not a row test but rather what was in this commit minus the last error case https://github.com/GoogleCloudPlatform/prometheus-engine/blob/f5edd69004e461dd2fa9dce874deb63521e7edea/pkg/e2e/fake_metrics_server_test.go#L163)

Reply:
I could also have multiple `CreateTimeSeries` in the row tests struct and loop over them too

Reply:
I prefer the latter actually